---
title: "QC Analysis of ICB IMmotion150"
author: "Nasim Bondar Sahebi"
date: "2024-06-13"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## load libraries

```{r libraries}
library(ggplot2)
library(dplyr)
library(MultiAssayExperiment) 
library(pheatmap)
library(DT)
library(tibble)
library(MultiAssayExperiment)
library(edgeR)
library(limma)
library(RColorBrewer)
library(ComplexHeatmap)
library(circlize)
```


## Data Loading and Preparation

**Data Overview**:

  - **Study Reference**: [PubMed ID PMC6721896](https://pubmed.ncbi.nlm.nih.gov/29867230/)
  - **Patient Count **: 326 ( 325 samples have Rna-seq ad 326 samples have WES)
  - **Multi-omices type**: Rna-seq and WES
  - **Treatment Details **:
  - **PD-1/PD-L1**: 110 patients
  - **IO+targeted**: 108 patients
  - **targeted**: 108 patients

Load multiassay .rds file, extract clinical, expression and annotations data; prepare gene expression data for analysis.

```{r load-data}
# Load your multiassay result and extract clinical data, expression data, and annotation

# Load mae obj
mae <- readRDS("~/BHK lab/ICB/ICB_IMmotion150/output data/ICB_IMmotion150.rds")

# Extract Clinical data
clin <- data.frame(colData(mae))

# mae has five assays: "expr_gene_tpm", "expr_gene_counts", "expr_isoform_tpm", "expr_isoform_counts" and "snv"

# Extract the 'expr_gene_tpm': TPM expression values for genes, normalized for length and sequencing depth.
expr <- assays(mae)[["expr_gene_tpm"]]  # dimension 61544 x 325
expr_counts <- assays(mae)[["expr_gene_counts"]] # dimension 61544 x 325
snv <- assays(mae)[["snv"]] # dimension 33039 x 326

# Extracting the annotation
annot_expr <- data.frame(rowData(mae@ExperimentList$expr_gene_tpm))
annot_snv <- data.frame(rowData(mae@ExperimentList$snv))

# Display first few rows of the dataset
DT::datatable(expr[1:8, 1:4])
DT::datatable(snv[1:8, 1:4])
DT::datatable(expr_counts[1:8, 1:4])

```

### Expression Preparation

This step involves subsetting our expression dataframe to include only protein-coding  as defined in the paper [PubMed ID PMC6721896](https://pubmed.ncbi.nlm.nih.gov/29867230/).

This prepares our dataset for the subsequent gene expression analysis.

```{r expr with gene of interest( proteincoding genes)}

## 1.Subset protein-coding genes for expr

# Step 1: Filter to include only protein-coding genes
annot_proteincoding <- annot_expr[annot_expr$gene_type == "protein_coding",]
expr <- expr[rownames(expr) %in% rownames(annot_proteincoding),] # dimension 19988 x 325

#set gene_id column for merging purpose 
expr$gene_id <- rownames(expr)

# Step 2: Merge to add gene names and remove gene_id column
expr <- merge(expr, annot_proteincoding[, c("gene_name", "gene_id")], by = "gene_id", all.x = TRUE)
expr$gene_id <- NULL  

# Step 3: Set gene names as row names for the expr dataframe
expr <- expr[!duplicated(expr$gene_name), ]
rownames(expr) <- expr$gene_name
expr$gene_name <- NULL

# Display expr
DT::datatable(expr[1:8, 1:4])

## 2.Same process for our count matrix
expr_counts <- assays(mae)[["expr_gene_counts"]]
expr_counts <- expr_counts[rownames(expr_counts) %in% rownames(annot_proteincoding),] # dimension 19988 x 325

expr_counts$gene_id <- rownames(expr_counts)

# Merge to add gene names and remove gene_id column
expr_counts <- merge(expr_counts, annot_proteincoding[, c("gene_name", "gene_id")], by = "gene_id", all.x = TRUE)
expr_counts$gene_id <- NULL
expr_counts <- expr_counts[!duplicated(expr_counts$gene_name), ]
rownames(expr_counts) <- expr_counts$gene_name
expr_counts$gene_name <- NULL

# Display our counts matrix
DT::datatable(expr_counts[1:8, 1:4])
```


### Gene Expression Analyses

This section outlines the analysis of gene expression using a heatmap, based on the methods detailed in the referenced study [PubMed ID PMC6721896](https://pubmed.ncbi.nlm.nih.gov/29867230/). The analysis includes 325 pretreatment tumor samples, exceeding the original study's 263 samples. 

**Objective**: To replicate and analyze the expression patterns of specific genes related to angiogenesis, immune function, and myeloid inflammation, comparing these patterns with those observed in Figure 2a of the referenced study.

**Steps**:
1. **Reverse Transformation**: Convert logged expression counts back to raw counts.
2. **Normalization and Transformation**:
   - Use edgeR’s TMM normalization to adjust for library size differences.
   - Filter out genes with low expression (below 0.25 CPM in less than 10% of samples).
   - Apply log2 transformation using limma’s `voom()` to stabilize variance across the samples.

**Gene of Interest**:
Specific genes analyzed include markers of angiogenesis (``VEGFA`, `KDR`, `ESM1`, `PECAM1`, `FLT1`, `ANGPTL4`, `CD34`), immune response (`CD8A`, `CD27`, `IFNG`, `GZMA`, `GZMB`, `PRF1`, `EOMES`, `CXCL9`, `CXCL10`, `CXCL11`, `CD274`, `CTLA4`, `FOXP3`, `TIGIT`), and myeloid inflammation (`IL6`, `CXCL1`, `CXCL2`, `CXCL3`, `CXCL8`, `PTGS2`, `PSMB8`, `PSMB9`, `TAP1`, `TAP2`, `IDO1`).

### Conclusion
Our extended dataset, processed using stringent normalization and transformation protocols, allows us to robustly compare and validate the reproducibility of gene expression profiles detailed in the original study. While differences in expression are noted, likely due to the increased sample size, our approach adheres to the same high quality standards. This comprehensive analysis may reveal additional biological subgroups, offering deeper insights into the complex dynamics of tumor biology.

```{r  heatmap, fig.width= 9 , fig.height= 7}
## 1. Define gene signatures

# Step : include only the genes of interest
# set Gene of Interest
genes_of_interest <- c(
  "VHL", "PBRM1", "VEGFA", "KDR", "ESM1", "PECAM1", "FLT1", "ANGPTL4", "CD34", "CD8A",
  "CD27", "IFNG", "GZMA", "GZMB", "PRF1", "EOMES", "CXCL9", "CXCL10", "CXCL11", "CD274",
  "CTLA4", "FOXP3", "TIGIT", "IDO1", "PSMB8", "PSMB9", "TAP1", "TAP2", "CXCL1", "CXCL2",
  "CXCL3", "CXCL8", "IL6", "PTGS2"
)

# Define gene categories
angiogenesis_genes <- c("VEGFA", "KDR", "ESM1", "PECAM1", "FLT1", "ANGPTL4", "CD34")
teff_genes <- c("CD8A", "CD27", "IFNG", "GZMA", "GZMB", "PRF1", "EOMES", "CXCL9", "CXCL10", "CXCL11", "CD274", "CTLA4", "FOXP3", "TIGIT")
myeloid_inflammation_genes <- c("IL6", "CXCL1", "CXCL2", "CXCL3", "CXCL8", "PTGS2", "PSMB8", "PSMB9", "TAP1", "TAP2", "IDO1")

## 2. Normalize data

# Convert log2(Counts + 1) back to Counts
expr_counts <- 2^expr_counts - 1

# Calculate normalization factors
dgList <- DGEList(counts = expr_counts)
dgList <- calcNormFactors(dgList, method="TMM")

# Filter out genes with low coverage
keep <- rowSums(cpm(dgList) > 0.25) >= (0.1 * ncol(dgList))
dgList <- dgList[keep,]

# Log2-transform using limma's voom
voomData <- voom(dgList, plot=TRUE)
expr_counts <- data.frame(voomData$E)  # log2-counts per million (log2-cpm) transformed data

# Transpose expr count 
expr_counts <- data.frame(t(expr_counts))
expr_counts$patientid <- rownames(expr_counts)

# Check for matching patient IDs
intersect_ids <- intersect(clin$patientid, expr_counts$patientid)
clin_data <- clin[clin$patientid %in% intersect_ids, ]

# Merge clinical data with expression data
data_merged <- merge(clin_data, expr_counts, by = "patientid", all.x = TRUE)

## 3. Set annotations 

# Prepare top annotations
top_annotation <- data.frame(
  PatientID = data_merged$patientid,
  PD_L1_IC_IHC_score = factor(data_merged$ICSCORE, levels = c("0", "1", "2", "3")),
  Sarcomatoid = factor(ifelse(data_merged$SarcomatoidFeatures == "Y", "Y", "N"), levels = c("N", "Y")),
  MSKCC = factor(data_merged$MSKCC, levels = c("Poor", "Intermediate", "Favorable")),
  Stage = factor(data_merged$stage, levels = c("I", "II", "III", "IV"))
)

rownames(top_annotation) <- top_annotation$PatientID
top_annotation$PatientID <- NULL
top_annotation <- na.omit(top_annotation)

## 5. Normalize Data Using Z-scores

filtered_expr_counts <- expr_counts[expr_counts$patientid %in% rownames(top_annotation), ] 
filtered_expr_counts$patientid <- NULL  #range -8.420901 13.446656
filtered_expr_counts <- filtered_expr_counts[, genes_of_interest]

# Z-score normalization for matrix of gene of interest
normalized_expr <- t(scale(t(filtered_expr_counts))) # range -2.83089  2.74211
normalized_expr <- normalized_expr [, genes_of_interest]

# Define the colors for the heatmap based on the actual range of the normalized data
expression_colors <- colorRamp2(c(-2, 0, 2), c("purple", "white", "yellow"))

# Create a single vector with all genes and their corresponding category
gene_categories <- c(rep("Angiogenesis", length(angiogenesis_genes)),
                     rep("Immune and Antigen Presentation", length(teff_genes)),
                     rep("Myeloid Inflammation", length(myeloid_inflammation_genes)))

# Combine the gene lists into a named vector
genes_of_interest <- c(angiogenesis_genes, teff_genes, myeloid_inflammation_genes)
names(gene_categories) <- genes_of_interest

# Create the annotation data frame
annotation_data <- data.frame(Category = gene_categories[rownames(t(normalized_expr))])
rownames(annotation_data) <- rownames(t(normalized_expr))

# Using pheatmap with simple annotations
pheatmap(as.matrix(t(normalized_expr)), 
         cluster_rows = FALSE, 
         cluster_cols = FALSE, 
         show_rownames = TRUE,
         show_colnames = FALSE,
         main = "Expression",
         annotation_col = top_annotation,  
         annotation_row = annotation_data,
         color = expression_colors)
```

### Heatmap with Gene Expression using gene level TPM

This heatmap displays gene expression data across various samples using `expr_gene_tpm` normalized with `log2(TPM+0.001)`. 


```{r heat map log-tpm , fig.width= 8, fig.height= 7}

#subset to gene of interest
expr <- expr[genes_of_interest,]

# Reset row names of clinical data
rownames(clin) <- NULL

# Prepare patient annotations
patient_annotation <- clin %>%
  dplyr::select(patientid, SarcomatoidFeatures, MSKCC, stage, VHL, PBRM1) %>%
  mutate(across(everything(), as.character)) %>%
  column_to_rownames(var = "patientid")

patient_annotation <- patient_annotation[colnames(expr), ]

# Generate the heatmap with annotations
pheatmap(expr, 
         cluster_rows = FALSE, 
         cluster_cols = FALSE, 
         show_rownames = TRUE,
         show_colnames = FALSE,
         annotation_col = patient_annotation)

```

