---
title: "QC Analysis of ICB_IMmotion150"
author: "Nasim Bondar Sahebi"
date: "2024-04-11"
output:
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## load libraries

```{r libraries}
library(ggplot2)
library(dplyr)
library(MultiAssayExperiment) 
library(pheatmap)
library(DT)
library(tibble)
library(MultiAssayExperiment)
library(edgeR)
library(limma)
library(RColorBrewer)
library(ComplexHeatmap)
library(circlize)
```


## Data Loading and Preparation

**Data Overview**:

  - **Study Reference**: [PubMed ID PMC6721896](https://pubmed.ncbi.nlm.nih.gov/29867230/)
  - **Patient Count **: 326 ( 325 samples have Rna-se ad 326 samples have WES)
  - ** Multi-omices type**: Rna-seq and WES
  - **Treatment Details **:
  - **PD-1/PD-L1**: 110 patients
  - **IO+targeted**: 108 patients
  - **targeted**: 108 patients

Load multiassay .rds file, extract clinical, expression and annotations data; prepare gene expression data for analysis.

```{r load-data}
# Load your multiassay result and extract clinical data, expression data, and annotation

# Load mae obj
mae <- readRDS("~/BHK lab/ICB/ICB_IMmotion150/output data/ICB_IMmotion150.rds")
# Extract Clinical data
clin <- data.frame(colData(mae))

# mae has five assays: "expr_gene_tpm", "expr_gene_counts", "expr_isoform_tpm", "expr_isoform_counts" and "snv"

# Extract the 'expr_gene_tpm': TPM expression values for genes, normalized for length and sequencing depth.
expr <- assays(mae)[["expr_gene_tpm"]]  # dimension 61544 x 325
expr_counts <- assays(mae)[["expr_gene_counts"]] # dimension 61544 x 325
snv <- assays(mae)[["snv"]] # dimension 33039 x 326

# Extracting the annotation
annot_expr <- data.frame(rowData(mae@ExperimentList$expr_gene_tpm))
annot_snv <- data.frame(rowData(mae@ExperimentList$snv))

# Display first few rows of the dataset
DT::datatable(expr[1:8, 1:4])
DT::datatable(snv[1:8, 1:4])
DT::datatable(expr_counts[1:8, 1:4])

```

### Expression Preparation

This step involves subsetting our expression dataframe to include only protein-coding genes and the Genes of Interest, as defined in the paper [PubMed ID PMC6721896](https://pubmed.ncbi.nlm.nih.gov/29867230/). Genes of Interest are specified as follows:

This prepares our dataset for the subsequent gene expression analysis.

```{r expr with gene of interest( proteincoding genes)}

## 1.Subset protein-coding genes for expr

# Step 1: Filter to include only protein-coding genes
annot_proteincoding <- annot_expr[annot_expr$gene_type == "protein_coding",]
expr <- expr[rownames(expr) %in% rownames(annot_proteincoding),] # dimension 19988 x 325

#set gene_id column for merging purpose 
expr$gene_id <- rownames(expr)

# Step 2: Merge to add gene names and remove gene_id column
expr <- merge(expr, annot_proteincoding[, c("gene_name", "gene_id")], by = "gene_id", all.x = TRUE)
expr$gene_id <- NULL  

# Step 3: Set gene names as row names for the expr dataframe
expr <- expr[!duplicated(expr$gene_name), ]
rownames(expr) <- expr$gene_name
expr$gene_name <- NULL

# Display expr
DT::datatable(expr)

## 2.Same process for our count matrix
expr_counts <- assays(mae)[["expr_gene_counts"]]
expr_counts <- expr_counts[rownames(expr_counts) %in% rownames(annot_proteincoding),] # dimension 19988 x 325

expr_counts$gene_id <- rownames(expr_counts)

# Merge to add gene names and remove gene_id column
expr_counts <- merge(expr_counts, annot_proteincoding[, c("gene_name", "gene_id")], by = "gene_id", all.x = TRUE)
expr_counts$gene_id <- NULL
expr_counts <- expr_counts[!duplicated(expr_counts$gene_name), ]
rownames(expr_counts) <- expr_counts$gene_name
expr_counts$gene_name <- NULL

# Display our counts matrix
DT::datatable(expr_counts)
```

### Gene Expression Analyses

Graphing a heatmap showing the expression of genes of interest (rows) as defined in the paper [PubMed ID PMC6721896](https://pubmed.ncbi.nlm.nih.gov/29867230/):




Sample annotations include:
- **PD-L1 IHC status** for tumor-infiltrating ICs.
- **Sarcomatoid Features**: Whether they are present or not.
- **MSKCC Score**: Classifying samples as Favorable, Intermediate, or Poor.
- **Tumor Stage**: Stage I-IV.
- **Tumor Mutation Burden (TMB)**: Number of mutations.
- **VHL and PBRM1**: Mutation status of these genes.
- **Signitures **
  - **Angio23**: `VEGFA`, `KDR`, `ESM1`, `PECAM1`, `ANGPTL4`, `CD34`
  - **Teff24**: `CD8A`, `EOMES`, `PRF1`, `IFNG`, `CD274`
  - **Myeloid Inflammation29–33**: `IL-6`, `CXCL1`, `CXCL2`, `CXCL3`, `CXCL8`, `PTGS2`

**Aim**: Comparing heatmap with Fig 2a from the referenced paper.

**Steps**:
1. **Normalization**: Convert `expr_counts` (log2(Counts + 1)) back to Counts.
2. Following the paper's normalization method: Normalize counts using edgeR’s normalization factors, filter out genes with low coverage (not reaching 0.25 CPM in at least one-tenth of samples), and apply log2-transformation using limma’s `voom()` function.

5. **expression is limited to gene of interest **:  Genes of Interest are specified as follows:

`VHL`, `PBRM1`, `VEGFA`, `KDR`, `ESM1`, `PECAM1`, `FLT1`, `ANGPTL4`, `CD34`, `CD8A`, `CD27`, `IFNG`, `GZMA`, `GZMB`, `PRF1`, `EOMES`, `CXCL9`, `CXCL10`, `CXCL11`, `CD274`, `CTLA4`, `FOXP3`, `TIGIT`, `IDO1`, `PSMB8`, `PSMB9`, `TAP1`, `TAP2`, `CXCL1`, `CXCL2`, `CXCL3`, `CXCL8`, `IL6`, `PTGS2`

**Conclusion**: The heatmap presents a detailed view of gene expression signatures, with clinical annotations and biological groupings, showing both similarities and differences compared to Fig 2a.


```{r  heatmap, fig.width= 9 , fig.height= 7}
## 1. Define gene signatures

# Step : include only the genes of interest
# set Gene of Interest
genes_of_interest <- c(
  "VHL", "PBRM1", "VEGFA", "KDR", "ESM1", "PECAM1", "FLT1", "ANGPTL4", "CD34", "CD8A",
  "CD27", "IFNG", "GZMA", "GZMB", "PRF1", "EOMES", "CXCL9", "CXCL10", "CXCL11", "CD274",
  "CTLA4", "FOXP3", "TIGIT", "IDO1", "PSMB8", "PSMB9", "TAP1", "TAP2", "CXCL1", "CXCL2",
  "CXCL3", "CXCL8", "IL6", "PTGS2"
)

# Define gene categories
angiogenesis_genes <- c("VEGFA", "KDR", "ESM1", "PECAM1", "FLT1", "ANGPTL4", "CD34")
teff_genes <- c("CD8A", "CD27", "IFNG", "GZMA", "GZMB", "PRF1", "EOMES", "CXCL9", "CXCL10", "CXCL11", "CD274", "CTLA4", "FOXP3", "TIGIT")
myeloid_inflammation_genes <- c("IL6", "CXCL1", "CXCL2", "CXCL3", "CXCL8", "PTGS2", "PSMB8", "PSMB9", "TAP1", "TAP2", "IDO1")

## 2. Normalize data

# Convert log2(Counts + 1) back to Counts
expr_counts <- 2^expr_counts - 1

# Calculate normalization factors
dgList <- DGEList(counts = expr_counts)
dgList <- calcNormFactors(dgList, method="TMM")

# Filter out genes with low coverage
keep <- rowSums(cpm(dgList) > 0.25) >= (0.1 * ncol(dgList))
dgList <- dgList[keep,]

# Log2-transform using limma's voom
voomData <- voom(dgList, plot=TRUE)
expr_counts <- data.frame(voomData$E)  # log2-counts per million (log2-cpm) transformed data

# Transpose expr count 
expr_counts <- data.frame(t(expr_counts))
expr_counts$patientid <- rownames(expr_counts)

# Check for matching patient IDs
intersect_ids <- intersect(clin$patientid, expr_counts$patientid)
clin_data <- clin[clin$patientid %in% intersect_ids, ]

# Merge clinical data with expression data
data_merged <- merge(clin_data, expr_counts, by = "patientid", all.x = TRUE)

## 3. Set annotations 

# Prepare top annotations
top_annotation <- data.frame(
  PatientID = data_merged$patientid,
  PD_L1_IC_IHC_score = factor(data_merged$ICSCORE, levels = c("0", "1", "2", "3")),
  Sarcomatoid = factor(ifelse(data_merged$SarcomatoidFeatures == "Y", "Y", "N"), levels = c("N", "Y")),
  MSKCC = factor(data_merged$MSKCC, levels = c("Poor", "Intermediate", "Favorable")),
  Stage = factor(data_merged$stage, levels = c("I", "II", "III", "IV"))
)

rownames(top_annotation) <- top_annotation$PatientID
top_annotation$PatientID <- NULL
top_annotation <- na.omit(top_annotation)

## 5. Normalize Data Using Z-scores

filtered_expr_counts <- expr_counts[expr_counts$patientid %in% rownames(top_annotation), ] 
filtered_expr_counts$patientid <- NULL  #range -8.420901 13.446656
filtered_expr_counts <- filtered_expr_counts[, genes_of_interest]

# Z-score normalization for matrix of gene of interest
normalized_expr <- t(scale(t(filtered_expr_counts))) # range -2.83089  2.74211
normalized_expr <- normalized_expr [, genes_of_interest]

# Define the colors for the heatmap based on the actual range of the normalized data
expression_colors <- colorRamp2(c(-2, 0, 2), c("purple", "white", "yellow"))

# Create a single vector with all genes and their corresponding category
gene_categories <- c(rep("Angiogenesis", length(angiogenesis_genes)),
                     rep("Immune and Antigen Presentation", length(teff_genes)),
                     rep("Myeloid Inflammation", length(myeloid_inflammation_genes)))

# Combine the gene lists into a named vector
genes_of_interest <- c(angiogenesis_genes, teff_genes, myeloid_inflammation_genes)
names(gene_categories) <- genes_of_interest

# Create the annotation data frame
annotation_data <- data.frame(Category = gene_categories[rownames(t(normalized_expr))])
rownames(annotation_data) <- rownames(t(normalized_expr))

# Using pheatmap with simple annotations
pheatmap(as.matrix(t(normalized_expr)), 
         cluster_rows = FALSE, 
         cluster_cols = FALSE, 
         show_rownames = TRUE,
         show_colnames = FALSE,
         main = "Expression",
         annotation_col = top_annotation,  
         annotation_row = annotation_data,
         color = expression_colors)
```

### Heatmap with Gene Expression using `log2(TPM+0.001)` Normalization

This heatmap displays gene expression data across various samples using a `log2(TPM+0.001)` normalization:

Comparing the present heatmap to Figure 2.a from the study [PMC6721896](https://pubmed.ncbi.nlm.nih.gov/29867230/):
1. **Normalization**: The study utilizes edgeR to normalize counts, transforming them to counts per million (CPM) before performing a z-score transformation. The current heatmap is immediately normalized to 'log2(TPM+0.001)'.
2. **Grouping and Sorting**: The paper's heatmap divides data into groups based on median signature scores for Angio, Teff, and Myeloid Inflammation, then sorts by these combinations. The present heatmap does not explicitly display this grouping and sorting.
3. **Clinical Annotations**: Both heatmaps contain clinical annotations, however the specific features and their relationship to gene expression may differ somewhat. The report adds additional annotations, such as PD-L1 IHC status and tumour mutation load.
4. **Expression Levels**: Both heatmaps use z-scores to show relative gene expression levels, allowing for a consistent comparison of expression patterns between samples.
5. **Visualization**: The present heatmap provides a simple view of gene expression signatures and clinical annotations, however the grouping and annotation details change somewhat from the relevant research.

In conclusion, both heatmaps give full insights into gene expression and clinical correlations, but they differ in normalization, grouping, and particular annotations, resulting in minor changes in patterns and context.




```{r heat map log-tpm , fig.width= 8 , fig.height= 5}

# second heatmap based on our processed data "expr", which represents gene expression data normalized to log2(TPM+0.001)

# Define gene signatures
angiogenesis_genes <- c("VEGFA", "KDR", "ESM1", "PECAM1", "ANGPTL4", "CD34")
teff_genes <- c("CD8A", "EOMES", "PRF1", "IFNG", "CD274")
myeloid_inflammation_genes <- c("IL6", "CXCL1", "CXCL2", "CXCL3", "CXCL8", "PTGS2")

# Calculate z-scores for the expression data
expr_z_scores <- t(scale(t(expr[, sapply(expr, is.numeric)])))

# Reset row names of clinical data
rownames(clin) <- NULL

# Prepare patient annotations
patient_annotation <- clin %>%
  dplyr::select(patientid, SarcomatoidFeatures, MSKCC, stage, VHL, PBRM1) %>%
  mutate(across(everything(), as.character)) %>%
  column_to_rownames(var = "patientid")

patient_annotation <- patient_annotation[colnames(expr_z_scores), ]

# Generate the heatmap with annotations
pheatmap(expr_z_scores, 
         cluster_rows = FALSE, 
         cluster_cols = FALSE, 
         show_rownames = TRUE,
         show_colnames = FALSE,
         annotation_col = patient_annotation)





```

